Index: src\states\adventure.c
===================================================================
--- src\states\adventure.c
+++ src\states\adventure.c
@@ -15,8 +15,9 @@
 #include "events.h"
 #include "rand.h"
 #include "vm.h"
 #include "math.h"
+#include "meta_tiles.h"
 #include "continuous_scene.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
@@ -235,10 +236,10 @@
 inline void adv_restore_default_anim_state(void)
 {
     if (adv_anim_dirty) {
         load_animations(PLAYER.sprite.ptr, PLAYER.sprite.bank, 0, PLAYER.animations);
-        actor_reset_anim(&PLAYER);  
-        adv_anim_dirty = FALSE;             
+        actor_reset_anim(&PLAYER);
+        adv_anim_dirty = FALSE;
     }
 }
 
 #endif
@@ -350,9 +351,9 @@
             case RUN_STATE: {
                 state_exit_run();
                 break;
             }
-#endif              
+#endif
 #ifdef FEAT_ADVENTURE_DASH
             case DASH_STATE: {
                 state_exit_dash();
                 break;
@@ -391,9 +392,9 @@
             case RUN_STATE: {
                 state_enter_run();
                 break;
             }
-#endif            
+#endif
 #ifdef FEAT_ADVENTURE_DASH
             case DASH_STATE: {
                 state_enter_dash();
                 break;
@@ -416,9 +417,9 @@
                 state_enter_push();
                 break;
             }
 #endif
-        }  
+        }
     }
 
 
     // STATE MACHINE =======================================================
@@ -473,9 +474,9 @@
             camera_deadzone_x--;
         } else if (camera_deadzone_x < adv_camera_deadzone_x) {
             camera_deadzone_x++;
         }
-        
+
         if (camera_deadzone_y > adv_camera_deadzone_y) {
             camera_deadzone_y--;
         } else if (camera_deadzone_y < adv_camera_deadzone_y) {
             camera_deadzone_y++;
@@ -484,9 +485,8 @@
             camera_deadzone_y == adv_camera_deadzone_y) {
             adv_camera_transitioning = FALSE;
         }
     }
-
 }
 
 static void handle_dir_input(void) {
     WORD target_x = 0;
@@ -496,9 +496,9 @@
     const WORD max_vel = (adv_state == RUN_STATE) ? adv_run_vel : adv_walk_vel;
     const WORD base_acc = (adv_state == RUN_STATE) ? adv_run_acc : adv_walk_acc;
 #else
     const WORD max_vel = adv_walk_vel;
-    const WORD base_acc = adv_walk_acc;    
+    const WORD base_acc = adv_walk_acc;
 #endif
 
 #if INPUT_ADVENTURE_MOVE_TYPE == MOVE_TYPE_8_WAY
     // ---------------------- 8-WAY ----------------------
@@ -514,10 +514,10 @@
     if (down)  target_y =  max_vel;
 
     if (target_x && target_y) {
         // Approximate diagonal movement normalization
-        target_x = (target_x >> 1) + (target_x >> 2);  
-        target_y = (target_y >> 1) + (target_y >> 2);  
+        target_x = (target_x >> 1) + (target_x >> 2);
+        target_y = (target_y >> 1) + (target_y >> 2);
     }
 #if INPUT_ADVENTURE_DIRECTION_TYPE == DIRECTION_TYPE_4_WAY
     // Update facing priority axis
     if ((left | right) && !(up | down)) {
@@ -535,19 +535,19 @@
         if (facing_priority_axis == DIR_PRIORITY_HORIZONTAL) {
             facing_dir = left ? DIR_LEFT : DIR_RIGHT;
         } else {
             facing_dir = up ? DIR_UP : DIR_DOWN;
-        }	
+        }
 #else
-	    facing_dir = left ? DIR_LEFT : DIR_RIGHT;
+        facing_dir = left ? DIR_LEFT : DIR_RIGHT;
 #endif
     } else if (left) {
         facing_dir = DIR_LEFT;
     } else if (right) {
         facing_dir = DIR_RIGHT;
-    } 
+    }
 #if INPUT_ADVENTURE_DIRECTION_TYPE != DIRECTION_TYPE_HORIZONTAL
-	else if (up) {
+    else if (up) {
         facing_dir = DIR_UP;
     } else if (down) {
         facing_dir = DIR_DOWN;
     }
@@ -568,9 +568,9 @@
         target_y =  max_vel;
         facing_dir = DIR_DOWN;
     }
 #else
-	if (INPUT_RECENT_LEFT) {
+    if (INPUT_RECENT_LEFT) {
         target_x = -max_vel;
         facing_dir = DIR_LEFT;
     } else if (INPUT_RECENT_RIGHT) {
         target_x =  max_vel;
@@ -657,9 +657,12 @@
                                 PLAYER.pos.y = target_y;
                             }
                         }
                     }
-                } 
+                }
+#ifdef ENABLE_ADV_RIGHT_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_RIGHT);
+#endif
             }
         } else if (delta.x < 0) {
             UBYTE tile_x = SUBPX_TO_TILE(new_x + PLAYER.bounds.left);
             UBYTE tile = tile_col_test_range_y(COLLISION_RIGHT, tile_x, tile_start, tile_end);
@@ -686,8 +689,11 @@
                             }
                         }
                     }
                 }
+#ifdef ENABLE_ADV_LEFT_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_LEFT);
+#endif
             }
         }
 
     finally_update_x:
@@ -719,27 +725,30 @@
                         if (PLAYER.pos.x < target_x + MAX_CORNER_PUSH_DISTANCE) {
                             adv_attached_actor = NULL;
                             PLAYER.pos.x -= VEL_TO_SUBPX(adv_walk_vel);
                             if (PLAYER.pos.x < target_x) {
-                                PLAYER.pos.x = target_x;   
+                                PLAYER.pos.x = target_x;
                             }
-                        }                            
+                        }
                     } else if (tile & COLLISION_RIGHT) {
                         UWORD target_x = TILE_TO_SUBPX(tile_hit_x + 1) + 1 - EXCLUSIVE_OFFSET(PLAYER.bounds.left);
                         if (PLAYER.pos.x > target_x - MAX_CORNER_PUSH_DISTANCE) {
                             adv_attached_actor = NULL;
                             PLAYER.pos.x += VEL_TO_SUBPX(adv_walk_vel);
                             if (PLAYER.pos.x > target_x) {
-                                PLAYER.pos.x = target_x;   
+                                PLAYER.pos.x = target_x;
                             }
                         }
-                    }                        
+                    }
                 }
+#ifdef ENABLE_ADV_DOWN_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_DOWN);
+#endif
             }
         } else if (delta.y < 0) {
             UBYTE tile_y = SUBPX_TO_TILE(new_y + PLAYER.bounds.top);
             UBYTE tile = tile_col_test_range_x(COLLISION_BOTTOM, tile_y, tile_start, tile_end);
-            if (tile) {            
+            if (tile) {
                 new_y = TILE_TO_SUBPX(tile_y + 1) - PLAYER.bounds.top;
                 adv_vel_y = 0;
                 if (tile & COLLISION_SLOPE && delta.x == 0) {
                     if (tile & COLLISION_LEFT) {
@@ -747,22 +756,25 @@
                         if (PLAYER.pos.x < target_x + MAX_CORNER_PUSH_DISTANCE) {
                             adv_attached_actor = NULL;
                             PLAYER.pos.x -= VEL_TO_SUBPX(adv_walk_vel);
                             if (PLAYER.pos.x < target_x) {
-                                PLAYER.pos.x = target_x;   
-                            }  
-                        }                          
+                                PLAYER.pos.x = target_x;
+                            }
+                        }
                     } else if (tile & COLLISION_RIGHT) {
                         UWORD target_x = TILE_TO_SUBPX(tile_hit_x + 1) + 1 - EXCLUSIVE_OFFSET(PLAYER.bounds.left);
                         if (PLAYER.pos.x > target_x - MAX_CORNER_PUSH_DISTANCE) {
                             adv_attached_actor = NULL;
                             PLAYER.pos.x += VEL_TO_SUBPX(adv_walk_vel);
                             if (PLAYER.pos.x > target_x) {
-                                PLAYER.pos.x = target_x;   
+                                PLAYER.pos.x = target_x;
                             }
                         }
                     }
                 }
+#ifdef ENABLE_ADV_UP_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_UP);
+#endif
             }
         }
 
     finally_update_y:
@@ -805,9 +817,9 @@
 
         if (adv_attached_actor != NULL) {
             hit_actor = adv_attached_actor;
             adv_attached_prev_x = hit_actor->pos.x;
-            adv_attached_prev_y = hit_actor->pos.y; 
+            adv_attached_prev_y = hit_actor->pos.y;
         } else {
             hit_actor = initial_hit_actor;
             adv_attached_actor = NULL;
             collision_dir = DIR_NONE;
@@ -825,20 +837,23 @@
             if (hit_actor && !(hit_actor->collision_group & COLLISION_GROUP_MASK) && hit_actor->script.bank) {
                 actor_set_dir(hit_actor, FLIPPED_DIR(PLAYER.dir), FALSE);
                 script_execute(hit_actor->script.bank, hit_actor->script.ptr, 0, 1, 0);
             }
-        }        
+        }
     }
-
+    
     //Check scene transition
     if (check_transition_to_scene_collision()) {
         return;
     }
 
     if (mask & COL_CHECK_TRIGGERS)
     {
         trigger_activate_at_intersection(&PLAYER.bounds, &PLAYER.pos, FALSE);
-    }    
+#ifdef ENABLE_ADV_ENTER_METATILE
+        metatile_overlap_at_intersection(&PLAYER.bounds, &PLAYER.pos);
+#endif
+    }
 }
 
 
 void adv_callback_attach(SCRIPT_CTX *THIS) OLDCALL BANKED
@@ -1016,9 +1031,9 @@
         WORD delta_mp_x = adv_attached_actor->pos.x - adv_attached_prev_x;
         WORD delta_mp_y = adv_attached_actor->pos.y - adv_attached_prev_y;
         adv_attached_prev_x = adv_attached_actor->pos.x;
         adv_attached_prev_y = adv_attached_actor->pos.y;
-        
+
         if (collision_dir == DIR_DOWN && delta_mp_y > adv_vel_y) {
             adv_vel_y = MAX(0, DELTA_TO_VEL(delta_mp_y));
         } else if (collision_dir == DIR_UP && delta_mp_y < adv_vel_y) {
             adv_vel_y = MIN(0, DELTA_TO_VEL(delta_mp_y));
@@ -1044,9 +1059,9 @@
     if (joy & INPUT_DPAD && prev_x == PLAYER.pos.x && prev_y == PLAYER.pos.y) {
         if (adv_push_timer == adv_push_delay_frames) {
             adv_next_state = PUSH_STATE;
             return;
-        }                
+        }
         adv_push_timer++;
     } else {
         adv_push_timer = 0;
     }
@@ -1075,9 +1090,9 @@
         actor_set_dir(&PLAYER, facing_dir, TRUE);
     } else {
         actor_set_anim_idle(&PLAYER);
     }
-    
+
 }
 
 // DASH_STATE
 
@@ -1090,9 +1105,9 @@
         return;
     }
 #ifdef ADVENTURE_DASH_ANIM
     adv_set_player_anim_state(ADVENTURE_DASH_ANIM);
-#elif ADVENTURE_ANIM_OVERRIDES_SET  
+#elif ADVENTURE_ANIM_OVERRIDES_SET
     adv_restore_default_anim_state();
 #endif
     adv_callback_execute(DASH_INIT);
 }
@@ -1246,9 +1261,9 @@
 
 static void state_update_push(void)
 {
     handle_dir_input();
-  
+
     delta.x = VEL_TO_SUBPX(adv_vel_x);
     delta.y = VEL_TO_SUBPX(adv_vel_y);
 
     UWORD prev_x = PLAYER.pos.x;
