Index: src\core\continuous_scene.c
===================================================================
--- src\core\continuous_scene.c
+++ src\core\continuous_scene.c
@@ -93,9 +93,9 @@
     actor_t *actor = actors_active_tail;
     while (actor) {
         if (actor != &PLAYER){
             SET_FLAG(actor->flags, ACTOR_FLAG_HIDDEN);
-        }		
+        }
         actor = actor->prev;
     }
     UBYTE tmp_show_actors_on_overlay = show_actors_on_overlay;
     show_actors_on_overlay = TRUE;
@@ -139,19 +139,19 @@
                 transitioning_player_pos_x = PLAYER.pos.x;
             }
         break;
     }
-    // kill all threads, but don't clear variables 
+    // kill all threads, but don't clear variables
     script_runner_init(FALSE);
     // reset timers on scene change
     timers_init(FALSE);
     // reset input events on scene change
     events_init(FALSE);
     // reset music events
     music_init_events(FALSE);
-    
+
     load_scene(continuous_scene->scene.ptr, continuous_scene->scene.bank, TRUE);
-    
+
     switch (direction){
         case DIRECTION_LEFT:
             camera_x += image_width_subpx;
             scroll_x += (image_width + 8);
@@ -189,9 +189,9 @@
     uint8_t direction = *(uint8_t *)VM_REF_TO_PTR(FN_ARG2);
     int8_t offset = *(int8_t *)VM_REF_TO_PTR(FN_ARG3);
     far_ptr_t background_col_ptr;
     ReadBankedFarPtr(&background_col_ptr, (void *)&(scene_ptr->background), scene_bank);
-	background_t bkg;
+    background_t bkg;
     MemcpyBanked(&bkg, background_col_ptr.ptr, sizeof(bkg), background_col_ptr.bank);
     ReadBankedFarPtr(&background_col_ptr, (void *)&(scene_ptr->collisions), scene_bank);
     continuous_scene_enabled |= DIRECTION_BIT(direction);
     continuous_scenes[direction].scene.bank = scene_bank;
@@ -208,9 +208,9 @@
     scroll_reset();
 }
 
 void remove_continuous_scene(SCRIPT_CTX * THIS) OLDCALL BANKED {
-    uint8_t direction = *(uint8_t *)VM_REF_TO_PTR(FN_ARG0);    
+    uint8_t direction = *(uint8_t *)VM_REF_TO_PTR(FN_ARG0);
     continuous_scene_enabled &= ~DIRECTION_BIT(direction);
     continuous_scenes[direction].scene.bank = 0;
     continuous_scenes[direction].scene.ptr = NULL;
     continuous_scenes[direction].tilemap.bank = 0;
@@ -228,15 +228,15 @@
 void load_scene_connections(SCRIPT_CTX * THIS) OLDCALL BANKED {
     uint8_t scene_connection_bank = *(uint8_t *) VM_REF_TO_PTR(FN_ARG0);
     const scene_connection_t* scene_connection_ptr = *(scene_connection_t**)VM_REF_TO_PTR(FN_ARG1);
     uint16_t connections_index = *(uint16_t *)VM_REF_TO_PTR(FN_ARG2) << 3; // multiply by number of directions which is 8
-    scene_connection_t connection; 
-    far_ptr_t background_col_ptr;   
+    scene_connection_t connection;
+    far_ptr_t background_col_ptr;
     background_t bkg;
-    for (uint8_t i = 0; i < 8; i++){        
+    for (uint8_t i = 0; i < 8; i++){
         MemcpyBanked(&connection, (scene_connection_ptr + (connections_index + (UWORD)i)), sizeof(scene_connection_t), scene_connection_bank);
-        if (connection.scene.ptr && connection.scene.bank){  
-            ReadBankedFarPtr(&background_col_ptr, (void *)&(((const scene_t *)connection.scene.ptr)->background), connection.scene.bank);	        
+        if (connection.scene.ptr && connection.scene.bank){
+            ReadBankedFarPtr(&background_col_ptr, (void *)&(((const scene_t *)connection.scene.ptr)->background), connection.scene.bank);
             MemcpyBanked(&bkg, background_col_ptr.ptr, sizeof(bkg), background_col_ptr.bank);
             ReadBankedFarPtr(&background_col_ptr, (void *)&(((const scene_t *)connection.scene.ptr)->collisions), connection.scene.bank);
             continuous_scene_enabled |= DIRECTION_BIT(i);
             continuous_scenes[i].scene.bank = connection.scene.bank;
