Index: src\states\platform.c
===================================================================
--- src\states\platform.c
+++ src\states\platform.c
@@ -15,8 +15,9 @@
 #include "scroll.h"
 #include "trigger.h"
 #include "vm.h"
 #include "macro.h"
+#include "meta_tiles.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
 
@@ -119,13 +120,16 @@
 #define COLLISION_SLOPE_45_RIGHT      COLLISION_SLOPE_45
 #define COLLISION_SLOPE_225_RIGHT_BOT COLLISION_SLOPE_225_BOT
 #define COLLISION_SLOPE_225_RIGHT_TOP COLLISION_SLOPE_225_TOP
 #define COLLISION_SLOPE_45_LEFT       (COLLISION_SLOPE_LEFT | COLLISION_SLOPE_45)
-#define COLLISION_SLOPE_225_LEFT_BOT  (COLLISION_SLOPE_LEFT | COLLISION_SLOPE_225_BOT) 
+#define COLLISION_SLOPE_225_LEFT_BOT  (COLLISION_SLOPE_LEFT | COLLISION_SLOPE_225_BOT)
 #define COLLISION_SLOPE_225_LEFT_TOP  (COLLISION_SLOPE_LEFT | COLLISION_SLOPE_225_TOP)
 #define COLLISION_SLOPE_ANY           (COLLISION_SLOPE_45 | COLLISION_SLOPE_225_BOT | COLLISION_SLOPE_225_TOP)
 #define COLLISION_SLOPE               0x70u
 
+#define ENABLE_PLAT_HORIZONTAL_COLLISION_METATILE \
+    defined(ENABLE_PLAT_LEFT_COLLISION_METATILE) || \
+    defined(ENABLE_PLAT_RIGHT_COLLISION_METATILE)
 
 #define PLATFORM_ANIM_OVERRIDES_SET \
     defined(PLATFORM_FALL_ANIM) || \
     (defined(FEAT_PLATFORM_FLOAT) && defined(PLATFORM_FLOAT_ANIM)) || \
@@ -395,9 +399,9 @@
     return INPUT_PRESSED(INPUT_PLATFORM_INTERACT);
 #elif INPUT_PLATFORM_DASH == DASH_INPUT_A
     return INPUT_PRESSED(INPUT_A);
 #elif INPUT_PLATFORM_DASH == DASH_INPUT_B
-    return INPUT_PRESSED(INPUT_B);    
+    return INPUT_PRESSED(INPUT_B);
 #elif INPUT_PLATFORM_DASH == DASH_INPUT_DOUBLE_TAP
     // Double-Tap Dash
     if (INPUT_PRESSED(INPUT_LEFT))
     {
@@ -420,9 +424,9 @@
         {
             plat_tap_timer = DOUBLE_TAP_WINDOW;
         }
     }
-    return FALSE;    
+    return FALSE;
 #elif INPUT_PLATFORM_DASH == DASH_INPUT_DOWN_JUMP
     return (INPUT_PRESSED(INPUT_DOWN) && INPUT_PLATFORM_JUMP) || (INPUT_DOWN && INPUT_PRESSED(INPUT_PLATFORM_JUMP));
 #else
     return FALSE;
@@ -443,58 +447,58 @@
 inline void plat_restore_default_anim_state(void)
 {
     if (plat_anim_dirty) {
         load_animations(PLAYER.sprite.ptr, PLAYER.sprite.bank, 0, PLAYER.animations);
-        actor_reset_anim(&PLAYER);  
-        plat_anim_dirty = FALSE;             
+        actor_reset_anim(&PLAYER);
+        plat_anim_dirty = FALSE;
     }
 }
 
 #endif
 
-static void state_enter_fall(void);      
-static void state_exit_fall(void);      
+static void state_enter_fall(void);
+static void state_exit_fall(void);
 static void state_update_fall(void);
-static void state_enter_ground(void);    
-static void state_exit_ground(void);    
+static void state_enter_ground(void);
+static void state_exit_ground(void);
 static void state_update_ground(void);
 #ifdef FEAT_PLATFORM_JUMP
-static void state_enter_jump(void);      
-static void state_exit_jump(void);      
+static void state_enter_jump(void);
+static void state_exit_jump(void);
 static void state_update_jump(void);
 #endif
 #ifdef FEAT_PLATFORM_DASH
-static void state_enter_dash(void);      
-static void state_exit_dash(void);      
+static void state_enter_dash(void);
+static void state_exit_dash(void);
 static void state_update_dash(void);
 #endif
 #ifdef FEAT_PLATFORM_LADDERS
-static void state_enter_ladder(void);    
-static void state_exit_ladder(void);    
+static void state_enter_ladder(void);
+static void state_exit_ladder(void);
 static void state_update_ladder(void);
 #endif
 #ifdef FEAT_PLATFORM_WALL_JUMP
-static void state_enter_wall(void);      
-static void state_exit_wall(void);      
+static void state_enter_wall(void);
+static void state_exit_wall(void);
 static void state_update_wall(void);
 #endif
 #ifdef FEAT_PLATFORM_KNOCKBACK
-static void state_enter_knockback(void); 
-static void state_exit_knockback(void); 
+static void state_enter_knockback(void);
+static void state_exit_knockback(void);
 static void state_update_knockback(void);
 #endif
 #ifdef FEAT_PLATFORM_BLANK
-static void state_enter_blank(void);     
-static void state_exit_blank(void);     
+static void state_enter_blank(void);
+static void state_exit_blank(void);
 static void state_update_blank(void);
 #endif
 #ifdef FEAT_PLATFORM_RUN
-static void state_enter_run(void);       
-static void state_exit_run(void);       
+static void state_enter_run(void);
+static void state_exit_run(void);
 #endif
 #ifdef FEAT_PLATFORM_FLOAT
-static void state_enter_float(void);     
-static void state_exit_float(void);     
+static void state_enter_float(void);
+static void state_exit_float(void);
 #endif
 
 // End of Function Definitions ------------------------------------------------
 
@@ -600,9 +604,9 @@
             state_exit_float();
             break;
         }
 #endif
-#ifdef FEAT_PLATFORM_JUMP        
+#ifdef FEAT_PLATFORM_JUMP
         case JUMP_STATE: {
             state_exit_jump();
             break;
         }
@@ -614,10 +618,10 @@
 #ifdef FEAT_PLATFORM_RUN
         case RUN_STATE: {
             state_exit_run();
             break;
-        }  
-#endif      
+        }
+#endif
 #ifdef FEAT_PLATFORM_LADDERS
         case LADDER_STATE: {
             state_exit_ladder();
             break;
@@ -666,9 +670,9 @@
             break;
         }
 #endif
 
-#ifdef FEAT_PLATFORM_JUMP        
+#ifdef FEAT_PLATFORM_JUMP
         case JUMP_STATE: {
             state_enter_jump();
             break;
         }
@@ -680,9 +684,9 @@
 #ifdef FEAT_PLATFORM_RUN
         case RUN_STATE: {
             state_enter_run();
             break;
-        }  
+        }
 #endif
 #ifdef FEAT_PLATFORM_LADDERS
         case LADDER_STATE: {
             state_enter_ladder();
@@ -934,10 +938,10 @@
         }
         if (col) {
             plat_next_state = FALL_STATE;
             return;
-        }     
-#endif        
+        }
+#endif
     }
 
     plat_is_actor_attached = FALSE;
     plat_camera_deadzone_x = camera_deadzone_x;
@@ -1141,9 +1145,9 @@
             bounds_edge = PLAYER.bounds.left;
         }
 
         UBYTE tile_x = SUBPX_TO_TILE(new_x + bounds_edge);
-        
+
         UBYTE tile = tile_col_test_range_y(hit_flag, tile_x, tile_y_start, tile_y_end);
 
         if (tile)
         {
@@ -1159,9 +1163,9 @@
                 (IS_ON_SLOPE(prev_on_slope) &&
                 (IS_SLOPE_LEFT(prev_on_slope) != moving_right)))
             {
                 goto finally_update_x;
-            }               
+            }
 #endif
             if (moving_right)
             {
                 new_x = TILE_TO_SUBPX(tile_x) - bounds_edge;
@@ -1176,8 +1180,29 @@
             plat_last_wall_col = wall;
 #ifdef FEAT_PLATFORM_WALL_JUMP
             plat_coyote_timer = plat_coyote_frames + 1;
 #endif
+#if ENABLE_PLAT_HORIZONTAL_COLLISION_METATILE
+            if (moving_right){
+#ifdef ENABLE_PLAT_RIGHT_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_RIGHT);
+#endif
+            } else {
+#ifdef ENABLE_PLAT_LEFT_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_LEFT);
+#endif
+            }
+        } else {
+            if (moving_right){
+#ifdef ENABLE_PLAT_RIGHT_COLLISION_METATILE
+                reset_collision_cache(DIR_RIGHT);
+#endif
+            } else {
+#ifdef ENABLE_PLAT_LEFT_COLLISION_METATILE
+                reset_collision_cache(DIR_LEFT);
+#endif
+            }
+#endif
         }
 
     finally_update_x:
         PLAYER.pos.x = new_x;
@@ -1318,8 +1343,13 @@
 #ifdef FEAT_PLATFORM_DROP_THROUGH
                 plat_drop_timer = 0;
 #endif
                 plat_grounded = TRUE;
+#ifdef ENABLE_PLAT_DOWN_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_DOWN);
+            } else {
+                reset_collision_cache(DIR_DOWN);
+#endif
             }
         finally_update_y:
             if (plat_grounded && (plat_state != DASH_STATE)) {
                 plat_next_state = GROUND_STATE;
@@ -1350,8 +1380,13 @@
 #ifdef FEAT_PLATFORM_COYOTE_TIME
                 plat_coyote_timer = 0;
 #endif
                 plat_next_state = FALL_STATE;
+#ifdef ENABLE_PLAT_UP_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_UP);
+            } else {
+                reset_collision_cache(DIR_UP);
+#endif
             }
             PLAYER.pos.y = new_y;
         }
     }
@@ -1408,9 +1443,9 @@
                     else if (PLAYER.pos.x != hit_actor->pos.x)
                     {
                         const UBYTE moving_right = PLAYER.pos.x < hit_actor->pos.x;
 
-                        plat_delta_x = (hit_actor->pos.x - PLAYER.pos.x) + 
+                        plat_delta_x = (hit_actor->pos.x - PLAYER.pos.x) +
                             (moving_right
                                 ? (hit_actor->bounds.left - EXCLUSIVE_OFFSET(PLAYER.bounds.right))
                                 : (EXCLUSIVE_OFFSET(hit_actor->bounds.right) - PLAYER.bounds.left));
                         PLAYER.pos.x += CLAMP(plat_delta_x, -MAX_DELTA, MAX_DELTA);
@@ -1454,8 +1489,11 @@
 
     if (mask & COL_CHECK_TRIGGERS)
     {
         trigger_activate_at_intersection(&PLAYER.bounds, &PLAYER.pos, INPUT_PLATFORM_FORCE_TRIGGER);
+#ifdef ENABLE_PLAT_ENTER_METATILE
+        metatile_overlap_at_intersection(&PLAYER.bounds, &PLAYER.pos);
+#endif
     }
 }
 
 void plat_callback_attach(SCRIPT_CTX *THIS) OLDCALL BANKED
@@ -1498,12 +1536,12 @@
     plat_grounded = FALSE;
     plat_is_actor_attached = FALSE;
 #ifdef PLATFORM_FALL_ANIM
     plat_set_player_anim_state(PLATFORM_FALL_ANIM);
-#elif PLATFORM_ANIM_OVERRIDES_SET  
+#elif PLATFORM_ANIM_OVERRIDES_SET
     plat_restore_default_anim_state();
 #endif
-    plat_callback_execute(FALL_INIT); 
+    plat_callback_execute(FALL_INIT);
 }
 
 static void state_exit_fall(void) {
     plat_callback_execute(FALL_END);
@@ -1857,9 +1895,9 @@
     plat_jump_vel_per_frame = plat_jump_hold_vel;
     // Calculate jump boost value based on horizontal velocity
     if (plat_run_boost != 0) {
         UWORD abs_vel_x = MAX(plat_vel_x, -plat_vel_x);
-#ifdef FEAT_PLATFORM_RUN                
+#ifdef FEAT_PLATFORM_RUN
         UBYTE boost_ratio = abs_vel_x / (plat_run_vel >> 7);
 #else
         UBYTE boost_ratio = abs_vel_x / (plat_walk_vel >> 7);
 #endif
@@ -1892,9 +1930,9 @@
         WORD reduced_vel_y = plat_vel_y + plat_jump_reduction_vel;
         plat_vel_y = MIN(reduced_vel_y, 0);
 #endif
 
-        // Handle jump velocity overflow 
+        // Handle jump velocity overflow
         if (plat_vel_y > 0) {
             plat_vel_y = WORD_MIN;
         }
 
@@ -2027,9 +2065,9 @@
         return;
     }
 #ifdef PLATFORM_DASH_ANIM
     plat_set_player_anim_state(PLATFORM_DASH_ANIM);
-#elif PLATFORM_ANIM_OVERRIDES_SET  
+#elif PLATFORM_ANIM_OVERRIDES_SET
     plat_restore_default_anim_state();
 #endif
     plat_callback_execute(DASH_INIT);
 }
@@ -2311,9 +2349,9 @@
 #ifdef PLATFORM_LADDERS_ANIM
         actor_set_anim_idle(&PLAYER);
 #else
         actor_stop_anim(&PLAYER);
-#endif            
+#endif
     } else {
         actor_set_anim(&PLAYER, ANIM_CLIMB);
     }
 
@@ -2331,12 +2369,12 @@
     plat_jump_type = JUMP_TYPE_NONE;
     plat_run_stage = RUN_STAGE_NONE;
 #ifdef PLATFORM_WALL_SLIDE_ANIM
     plat_set_player_anim_state(PLATFORM_WALL_SLIDE_ANIM);
-#elif PLATFORM_ANIM_OVERRIDES_SET  
+#elif PLATFORM_ANIM_OVERRIDES_SET
     plat_restore_default_anim_state();
 #endif
-    plat_callback_execute(WALL_INIT);    
+    plat_callback_execute(WALL_INIT);
 }
 
 static void state_exit_wall(void) {
     plat_callback_execute(WALL_END);
@@ -2432,9 +2470,9 @@
     plat_knockback_timer = plat_knockback_frames;
     plat_drop_timer = 0;
 #ifdef PLATFORM_KNOCKBACK_ANIM
     plat_set_player_anim_state(PLATFORM_KNOCKBACK_ANIM);
-#elif PLATFORM_ANIM_OVERRIDES_SET  
+#elif PLATFORM_ANIM_OVERRIDES_SET
     plat_restore_default_anim_state();
 #endif
     plat_callback_execute(KNOCKBACK_INIT);
 }
@@ -2442,9 +2480,9 @@
 static void state_exit_knockback(void) {
     plat_vel_x = 0;
     plat_callback_execute(KNOCKBACK_END);
 }
-        
+
 static void state_update_knockback(void) {
     // Horizontal Movement --------------------------------------------
 
     // Horizontal deceleration
@@ -2454,9 +2492,9 @@
     plat_delta_x += VEL_TO_SUBPX(plat_vel_x);
 
     // Vertical Movement ----------------------------------------------
 
-    // Normal gravity        
+    // Normal gravity
     plat_vel_y += plat_grav;
     plat_vel_y = MIN(plat_vel_y, plat_max_fall_vel);
     plat_delta_y += VEL_TO_SUBPX(plat_vel_y);
 
@@ -2464,9 +2502,9 @@
 
     move_and_collide(COL_CHECK_ALL);
 
     // Counters -------------------------------------------------------
-    
+
     COUNTER_DECREMENT(plat_knockback_timer);
     if (plat_knockback_timer != 0) {
         plat_next_state = KNOCKBACK_STATE;
     }
@@ -2483,18 +2521,18 @@
     plat_run_stage = RUN_STAGE_NONE;
     plat_jump_type = JUMP_TYPE_NONE;
 #ifdef PLATFORM_BLANK_ANIM
     plat_set_player_anim_state(PLATFORM_BLANK_ANIM);
-#elif PLATFORM_ANIM_OVERRIDES_SET  
+#elif PLATFORM_ANIM_OVERRIDES_SET
     plat_restore_default_anim_state();
 #endif
-    plat_callback_execute(BLANK_INIT);  
+    plat_callback_execute(BLANK_INIT);
 }
 
 static void state_exit_blank(void) {
     plat_vel_x = 0;
     plat_vel_y = 0;
-    plat_callback_execute(BLANK_END);    
+    plat_callback_execute(BLANK_END);
 }
 
 static void state_update_blank(void) {
     // Vertical Movement ----------------------------------------------
